### ISSUE TITLE
[ENH] Add Tracking Error as a risk measure

### ISSUE THREAD

**Issue by Gilvir** [skfolio/skfolio#195]
**Is your feature request related to a problem? Please describe.**
Currently, tracking error can only be used as a constraint (max_tracking_error), not as an optimization objective. This makes it impossible to directly minimize tracking error, which is a common problem in index tracking and portfolio management.

**Describe the solution you'd like**
- Add TRACKING_ERROR as a new RiskMeasure option.
- Add benchmark_weights parameter to MeanRisk.
- Add the tracking_error_risk, which we calculate as the standard deviation of the active weights 
    (portfolio weights minus benchmark weights)

Example usage
```
model = MeanRisk(
    objective_function=ObjectiveFunction.MINIMIZE_RISK,
    risk_measure=RiskMeasure.TRACKING_ERROR,
    benchmark_weights=[0.25, 0.25, 0.25, 0.25]
)
model.fit(X_train)
```

**Describe alternatives you've considered**
- Using max_tracking_error with small values, dosent really minimize the error.
- Transforming to excess returns, only works ex post.

**Additional context**
Tracking error is widely used by institutional funds for index tracking and other optimizations.

**Comment by HugoDelatte** [comment]
Yes indeed, this issue is under my radar as I'm currently working on a new fundamental factor model feature and I'm playing around with the API to find the best design.
The solution would be to use the new scikit-learn API "metadata routing" / `fit_params` for the factor models.
I think we should keep `y` for the tracking error case as it really represent a "target".
I'm also playing with factor model `prefit` and `partial_fit` (but more for speed/online learning).

**Comment by tingiskhan** [comment]
I would really like this!

> For the tracking error minimization, I think it should be diffed the same way instead of a target weight: your target return is not necessarily a function of your investment universe (for example we could use a macro ts or an Index/ETF not in X) and also if we define the loss based on a target weights it will converge to the exact target weights (without additional constraints).

From what I gather, index/benchmark tracking managers only care about _ex-ante_ risk as opposed to ex-post risk. The resulting weights utilizing the time series approach would become highly dependent on the chosen lookback period, causing issues around index rebalancing, right? But I also do agree with your point. Would it therefore be possible to support both? One (albeit implicit) way of handling this would be to handle `y` differently depending on the its index:
- If it's of date time type, do RMSE.
- Else, do the one @Gilvir proposed.

And regarding the "metadata routing", would it result in something like below?

```python
problem = MeanRisk(...)

problem.fit(X, y, factor_returns=f)
```

**Comment by Gilvir** [comment]
Created a draft PR implementing ex-ante and ex-post tracking error using the `y` parameter for the benchmark target. Also added metadata routing to FactorModel. Feedback welcome!

**Comment by HugoDelatte** [comment]
@Gilvir very nice, I've replied on the PR

**Comment by dppalomar** [comment]
The topic of index tracking is very relevant to me. I see you were referring to my slides above. Note that more detailed information can be found in Chapter 13 of my book:

_Portfolio Optimization: Theory and Application_, Cambridge Univ. Press, 2025.
https://portfoliooptimizationbook.com/

Here is the link to Chapter 13 on index tracking:
https://bookdown.org/palomar/portfoliooptimizationbook/13-index-tracking.html

I wanted to highlight two points that I think could be easily incorporated in skfolio:

1. **Different tracking error measures**. The typical one is the $\ell_2$-norm tracking error (https://bookdown.org/palomar/portfoliooptimizationbook/13.3-sparse-index-tracking.html#tracking-error). However, it is straightforward to consider alternative (convex) error measures since skfolio is using cvxpy under the hood. Here is a list:
    + TE_2 (traditional tracking error using the $\ell_2$-norm): https://bookdown.org/palomar/portfoliooptimizationbook/13.3-sparse-index-tracking.html#tracking-error
    + TE_1 (tracking error using $\ell_1$-norm): https://bookdown.org/palomar/portfoliooptimizationbook/13.4-extensions-index-tracking.html#ell_1-norm-tracking-error
    + DR (downside risk): https://bookdown.org/palomar/portfoliooptimizationbook/13.4-extensions-index-tracking.html#downside-risk-error-measure (this measure could be based on either $\ell_2$-norm or $\ell_1$-norm)
    + Robust Huberized-TE_2: https://bookdown.org/palomar/portfoliooptimizationbook/13.4-extensions-index-tracking.html#robust-tracking-error-measures

Note that in Chapter 13 I develop simple algorithms based on the Majorization-Minimization (MM) framework, but this is not necessary under skfolio since it is using cvxpy under the hood and all the tracking errors are convex.

2. **Type of returns used in the definitition of the tracking error**. Again, people typically use the period-by-period returns. However, as explored in Chapter 13 (https://bookdown.org/palomar/portfoliooptimizationbook/13.3-sparse-index-tracking.html), there are many variations of returns one can use. There are four important types that arise from two different considerations:
    + cumulative vs. non-cumulative returns: roughly speaking maps into long-term vs. short-term tracking (of tracking of prices vs. returns), see https://bookdown.org/palomar/portfoliooptimizationbook/13.3-sparse-index-tracking.html#plain-vs.-cumulative-returns
    + fixed vs. time-varying portfolio: this is a new concept that was uncovered in my book for the first time, it turns out that it is straightforward to implement; see https://bookdown.org/palomar/portfoliooptimizationbook/13.3-sparse-index-tracking.html#fixed-vs.-time-varying-portfolio

Here is the summary of these four cases of returns: https://bookdown.org/palomar/portfoliooptimizationbook/13.3-sparse-index-tracking.html#summary-of-tracking-errors

If you need any help or need to discuss anything, I am happy to help.

@HugoDelatte @Gilvir @CarloNicolini @tingiskhan @8W9aG
